Support message content parts in chat completion requests

`request.messages.*.content` may contain a list of content parts, but currently, the service presumes only a string content: 

https://github.com/epam/ai-dial-analytics-realtime/blob/01f8f42dc8cbfb98c40e7890032d6a9d6b8a0c19/aidial_analytics_realtime/analytics.py#L83-L85

Therefore it fails with the error:

```
2024-12-06 17:11:40,382 [ERROR] - sequence item 17: expected str instance, list found
Traceback (most recent call last):
  File "/aidial_analytics_realtime/app.py", line 294, in on_log_messages
    await on_log_message(
  File "/aidial_analytics_realtime/app.py", line 242, in on_log_message
    await on_chat_completion_message(
  File "/aidial_analytics_realtime/app.py", line 134, in on_chat_completion_message
    await on_message(
  File "/aidial_analytics_realtime/analytics.py", line 254, in on_message
    point = make_point(
  File "/aidial_analytics_realtime/analytics.py", line 83, in make_point
    request_content = "\n".join(
TypeError: sequence item 17: expected str instance, list found
```

--- a/aidial_analytics_realtime/analytics.py
+++ b/aidial_analytics_realtime/analytics.py
@@ -2,11 +2,12 @@ from datetime import datetime
 from decimal import Decimal
 from enum import Enum
 from logging import Logger
-from typing import Awaitable, Callable, List
+from typing import Any, Awaitable, Callable, List
 from uuid import uuid4
 
 from influxdb_client import Point
 from langid.langid import LanguageIdentifier, model
+from typing_extensions import assert_never
 
 from aidial_analytics_realtime.rates import RatesCalculator
 from aidial_analytics_realtime.topic_model import TopicModel
@@ -19,45 +20,12 @@ class RequestType(Enum):
     EMBEDDING = 2
 
 
-def detect_lang(request, response, request_type):
-    if request_type == RequestType.CHAT_COMPLETION:
-        text = (
-            request["messages"][-1]["content"]
-            + "\n\n"
-            + response["choices"][0]["message"]["content"]
-        )
-    else:
-        text = (
-            request["input"]
-            if isinstance(request["input"], str)
-            else "\n\n".join(request["input"])
-        )
-
-    return detect_lang_by_text(text)
-
-
-def detect_lang_by_text(text):
-    try:
-        lang, prob = identifier.classify(text)
-
-        if prob > 0.998:
-            return lang
-
-        return "undefined"
-    except Exception:
-        return "undefined"
-
-
-def to_string(obj: str | None):
-    return obj if obj else "undefined"
-
-
-def build_execution_path(path: list | None):
-    return "undefined" if not path else "/".join(map(to_string, path))
-
-
 def get_chat_completion_request_contents(request: dict) -> List[str]:
     return [message["content"] for message in request["messages"]]
+
+
+def get_chat_completion_response_contents(response: dict) -> List[str]:
+    return [response["choices"][0]["message"]["content"]]
 
 
 def get_embeddings_request_contents(request: dict) -> List[str]:
@@ -66,6 +34,46 @@ def get_embeddings_request_contents(request: dict) -> List[str]:
         if isinstance(request["input"], str)
         else request["input"]
     )
+
+
+def detect_lang(
+    request: dict, response: dict, request_type: RequestType
+) -> str:
+    match request_type:
+        case RequestType.CHAT_COMPLETION:
+            request_contents = get_chat_completion_request_contents(request)
+            response_content = get_chat_completion_response_contents(response)
+            text = "\n\n".join(request_contents[-1:] + response_content)
+        case RequestType.EMBEDDING:
+            text = "\n\n".join(get_embeddings_request_contents(request))
+        case _:
+            assert_never(request_type)
+
+    return to_string(detect_lang_by_text(text))
+
+
+def detect_lang_by_text(text: str) -> str | None:
+    text = text.strip()
+
+    if not text:
+        return None
+
+    try:
+        lang, prob = identifier.classify(text)
+        if prob > 0.998:
+            return lang
+    except Exception:
+        pass
+
+    return None
+
+
+def to_string(obj: str | None) -> str:
+    return obj or "undefined"
+
+
+def build_execution_path(path: list | None):
+    return "undefined" if not path else "/".join(map(to_string, path))
 
 
 def make_point(
@@ -90,22 +98,28 @@ def make_point(
     topic = None
     response_content = ""
     request_content = ""
-    if request_type == RequestType.CHAT_COMPLETION:
-        response_content = response["choices"][0]["message"]["content"]
+    match request_type:
+        case RequestType.CHAT_COMPLETION:
+            response_contents = get_chat_completion_response_contents(response)
+            request_contents = get_chat_completion_request_contents(request)
 
-        request_contents = get_chat_completion_request_contents(request)
+            request_content = "\n".join(request_contents)
+            response_content = "\n".join(response_contents)
 
-        request_content = "\n".join(request_contents)
-        if chat_id:
-            topic = topic_model.get_topic_by_text(
-                "\n\n".join(request_contents + [response_content])
-            )
-    else:
-        request_contents = get_embeddings_request_contents(request)
+            if chat_id:
+                topic = topic_model.get_topic_by_text(
+                    "\n\n".join(request_contents + response_contents)
+                )
+        case RequestType.EMBEDDING:
+            request_contents = get_embeddings_request_contents(request)
 
-        request_content = "\n".join(request_contents)
-        if chat_id:
-            topic = topic_model.get_topic_by_text("\n\n".join(request_contents))
+            request_content = "\n".join(request_contents)
+            if chat_id:
+                topic = topic_model.get_topic_by_text(
+                    "\n\n".join(request_contents)
+                )
+        case _:
+            assert_never(request_type)
 
     price = Decimal(0)
     deployment_price = Decimal(0)
@@ -135,7 +149,7 @@ def make_point(
             (
                 "undefined"
                 if not trace
-                else to_string(trace.get("core_parent_span_id", None))
+                else to_string(trace.get("core_parent_span_id"))
             ),
         )
         .tag("project_id", project_id)
@@ -221,7 +235,7 @@ def make_rate_point(
 
 
 async def parse_usage_per_model(response: dict):
-    statistics = response.get("statistics", None)
+    statistics = response.get("statistics")
     if statistics is None:
         return []
 
@@ -293,7 +307,7 @@ async def on_message(
             request,
             response,
             type,
-            response.get("usage", None),
+            response.get("usage"),
             topic_model,
             rates_calculator,
             parent_deployment,

--- a/aidial_analytics_realtime/analytics.py
+++ b/aidial_analytics_realtime/analytics.py
@@ -2,7 +2,7 @@ from datetime import datetime
 from decimal import Decimal
 from enum import Enum
 from logging import Logger
-from typing import Awaitable, Callable
+from typing import Awaitable, Callable, List
 from uuid import uuid4
 
 from influxdb_client import Point
@@ -56,6 +56,18 @@ def build_execution_path(path: list | None):
     return "undefined" if not path else "/".join(map(to_string, path))
 
 
+def get_chat_completion_request_contents(request: dict) -> List[str]:
+    return [message["content"] for message in request["messages"]]
+
+
+def get_embeddings_request_contents(request: dict) -> List[str]:
+    return (
+        [request["input"]]
+        if isinstance(request["input"], str)
+        else request["input"]
+    )
+
+
 def make_point(
     deployment: str,
     model: str,
@@ -80,23 +92,20 @@ def make_point(
     request_content = ""
     if request_type == RequestType.CHAT_COMPLETION:
         response_content = response["choices"][0]["message"]["content"]
-        request_content = "\n".join(
-            [message["content"] for message in request["messages"]]
-        )
-        if chat_id:
-            topic = topic_model.get_topic(request["messages"], response_content)
-    else:
-        request_content = (
-            request["input"]
-            if isinstance(request["input"], str)
-            else "\n".join(request["input"])
-        )
+
+        request_contents = get_chat_completion_request_contents(request)
+
+        request_content = "\n".join(request_contents)
         if chat_id:
             topic = topic_model.get_topic_by_text(
-                request["input"]
-                if isinstance(request["input"], str)
-                else "\n\n".join(request["input"])
+                "\n\n".join(request_contents + [response_content])
             )
+    else:
+        request_contents = get_embeddings_request_contents(request)
+
+        request_content = "\n".join(request_contents)
+        if chat_id:
+            topic = topic_model.get_topic_by_text("\n\n".join(request_contents))
 
     price = Decimal(0)
     deployment_price = Decimal(0)

--- a/aidial_analytics_realtime/analytics.py
+++ b/aidial_analytics_realtime/analytics.py
@@ -2,13 +2,18 @@ from datetime import datetime
 from decimal import Decimal
 from enum import Enum
 from logging import Logger
-from typing import Any, Awaitable, Callable, List
+from typing import Awaitable, Callable
 from uuid import uuid4
 
 from influxdb_client import Point
 from langid.langid import LanguageIdentifier, model
 from typing_extensions import assert_never
 
+from aidial_analytics_realtime.dial import (
+    get_chat_completion_request_contents,
+    get_chat_completion_response_contents,
+    get_embeddings_request_contents,
+)
 from aidial_analytics_realtime.rates import RatesCalculator
 from aidial_analytics_realtime.topic_model import TopicModel
 
@@ -20,32 +25,20 @@ class RequestType(Enum):
     EMBEDDING = 2
 
 
-def get_chat_completion_request_contents(request: dict) -> List[str]:
-    return [message["content"] for message in request["messages"]]
-
-
-def get_chat_completion_response_contents(response: dict) -> List[str]:
-    return [response["choices"][0]["message"]["content"]]
-
-
-def get_embeddings_request_contents(request: dict) -> List[str]:
-    return (
-        [request["input"]]
-        if isinstance(request["input"], str)
-        else request["input"]
-    )
-
-
 def detect_lang(
-    request: dict, response: dict, request_type: RequestType
+    logger: Logger, request: dict, response: dict, request_type: RequestType
 ) -> str:
     match request_type:
         case RequestType.CHAT_COMPLETION:
-            request_contents = get_chat_completion_request_contents(request)
-            response_content = get_chat_completion_response_contents(response)
+            request_contents = get_chat_completion_request_contents(
+                logger, request
+            )
+            response_content = get_chat_completion_response_contents(
+                logger, response
+            )
             text = "\n\n".join(request_contents[-1:] + response_content)
         case RequestType.EMBEDDING:
-            text = "\n\n".join(get_embeddings_request_contents(request))
+            text = "\n\n".join(get_embeddings_request_contents(logger, request))
         case _:
             assert_never(request_type)
 
@@ -77,6 +70,7 @@ def build_execution_path(path: list | None):
 
 
 def make_point(
+    logger: Logger,
     deployment: str,
     model: str,
     project_id: str,
@@ -100,8 +94,12 @@ def make_point(
     request_content = ""
     match request_type:
         case RequestType.CHAT_COMPLETION:
-            response_contents = get_chat_completion_response_contents(response)
-            request_contents = get_chat_completion_request_contents(request)
+            response_contents = get_chat_completion_response_contents(
+                logger, response
+            )
+            request_contents = get_chat_completion_request_contents(
+                logger, request
+            )
 
             request_content = "\n".join(request_contents)
             response_content = "\n".join(response_contents)
@@ -111,7 +109,7 @@ def make_point(
                     "\n\n".join(request_contents + response_contents)
                 )
         case RequestType.EMBEDDING:
-            request_contents = get_embeddings_request_contents(request)
+            request_contents = get_embeddings_request_contents(logger, request)
 
             request_content = "\n".join(request_contents)
             if chat_id:
@@ -158,7 +156,7 @@ def make_point(
             (
                 "undefined"
                 if not chat_id
-                else detect_lang(request, response, request_type)
+                else detect_lang(logger, request, response, request_type)
             ),
         )
         .tag("upstream", to_string(upstream_url))
@@ -275,6 +273,7 @@ async def on_message(
     usage_per_model = await parse_usage_per_model(response)
     if token_usage is not None:
         point = make_point(
+            logger,
             deployment,
             model,
             project_id,
@@ -296,6 +295,7 @@ async def on_message(
         await influx_writer(point)
     elif len(usage_per_model) == 0:
         point = make_point(
+            logger,
             deployment,
             model,
             project_id,
@@ -317,6 +317,7 @@ async def on_message(
         await influx_writer(point)
     else:
         point = make_point(
+            logger,
             deployment,
             model,
             project_id,
@@ -339,6 +340,7 @@ async def on_message(
 
         for usage in usage_per_model:
             point = make_point(
+                logger,
                 deployment,
                 usage["model"],
                 project_id,

--- a/aidial_analytics_realtime/app.py
+++ b/aidial_analytics_realtime/app.py
@@ -291,8 +291,6 @@ async def on_log_messages(
     data = await request.json()
 
     statuses = []
-    any_errors = False
-
     for idx, item in enumerate(data):
         try:
             await on_log_message(
@@ -303,14 +301,13 @@ async def on_log_messages(
             )
         except Exception as e:
             logging.exception(f"Error processing message #{idx}")
-            any_errors = True
             statuses.append({"status": "error", "error": str(e)})
         else:
             statuses.append({"status": "success"})
 
-    status_code = 500 if any_errors else 200
-
-    return JSONResponse(content=statuses, status_code=status_code)
+    # Returning 200 code even if processing of some messages has failed,
+    # since the log broker that sends the messages may decide to retry the failed requests.
+    return JSONResponse(content=statuses, status_code=200)
 
 
 @app.get("/health")

--- a/aidial_analytics_realtime/app.py
+++ b/aidial_analytics_realtime/app.py
@@ -224,8 +224,7 @@ async def on_log_message(
     execution_path = message.get("execution_path", None)
     deployment = message.get("deployment", "")
 
-    match = re.search(RATE_PATTERN, uri)
-    if match:
+    if re.search(RATE_PATTERN, uri):
         await on_rate_message(
             deployment,
             project_id,
@@ -238,8 +237,7 @@ async def on_log_message(
             influx_writer,
         )
 
-    match = re.search(CHAT_COMPLETION_PATTERN, uri)
-    if match:
+    elif re.search(CHAT_COMPLETION_PATTERN, uri):
         await on_chat_completion_message(
             deployment,
             project_id,
@@ -259,8 +257,7 @@ async def on_log_message(
             execution_path,
         )
 
-    match = re.search(EMBEDDING_PATTERN, uri)
-    if match:
+    elif re.search(EMBEDDING_PATTERN, uri):
         await on_embedding_message(
             deployment,
             project_id,
@@ -279,6 +276,9 @@ async def on_log_message(
             trace,
             execution_path,
         )
+
+    else:
+        logger.warning(f"Unsupported message type: {uri!r}")
 
 
 @app.post("/data")

--- a/aidial_analytics_realtime/app.py
+++ b/aidial_analytics_realtime/app.py
@@ -5,6 +5,7 @@ from datetime import datetime
 
 import uvicorn
 from fastapi import Depends, FastAPI, Request
+from fastapi.responses import JSONResponse
 
 from aidial_analytics_realtime.analytics import (
     RequestType,
@@ -289,7 +290,10 @@ async def on_log_messages(
 ):
     data = await request.json()
 
-    for item in data:
+    statuses = []
+    any_errors = False
+
+    for idx, item in enumerate(data):
         try:
             await on_log_message(
                 json.loads(item["message"]),
@@ -298,7 +302,15 @@ async def on_log_messages(
                 rates_calculator,
             )
         except Exception as e:
-            logging.exception(e)
+            logging.exception(f"Error processing message #{idx}")
+            any_errors = True
+            statuses.append({"status": "error", "error": str(e)})
+        else:
+            statuses.append({"status": "success"})
+
+    status_code = 500 if any_errors else 200
+
+    return JSONResponse(content=statuses, status_code=status_code)
 
 
 @app.get("/health")

--- a/aidial_analytics_realtime/dial.py
+++ b/aidial_analytics_realtime/dial.py
@@ -1,0 +1,53 @@
+from logging import Logger
+from typing import List
+
+
+def get_chat_completion_request_contents(
+    logger: Logger, request: dict
+) -> List[str]:
+    return [
+        content
+        for message in request["messages"]
+        for content in _get_chat_completion_message_contents(logger, message)
+    ]
+
+
+def get_chat_completion_response_contents(
+    logger: Logger, response: dict
+) -> List[str]:
+    message = response["choices"][0]["message"]
+    return _get_chat_completion_message_contents(logger, message)
+
+
+def get_embeddings_request_contents(logger: Logger, request: dict) -> List[str]:
+    inp = request.get("input")
+
+    if isinstance(inp, str):
+        return [inp]
+    elif isinstance(inp, list):
+        return [i for i in inp if isinstance(i, str)]
+    else:
+        logger.warning(f"Unexpected type of embeddings input: {type(inp)}")
+        return []
+
+
+def _get_chat_completion_message_contents(
+    logger: Logger, message: dict
+) -> List[str]:
+    content = message.get("content")
+    if content is None:
+        return []
+    elif isinstance(content, str):
+        return [content]
+    elif isinstance(content, list):
+        ret: List[str] = []
+        for content_part in content:
+            if isinstance(content_part, dict):
+                if content_part.get("type") == "text" and (
+                    text := content_part.get("content")
+                ):
+                    ret.extend(text)
+        return ret
+    else:
+        logger.warning(f"Unexpected message content type: {type(content)}")
+        return []

--- a/aidial_analytics_realtime/topic_model.py
+++ b/aidial_analytics_realtime/topic_model.py
@@ -20,14 +20,6 @@ class TopicModel:
         )
         self.model.transform(["test"])  # Make sure the model is loaded
 
-    def get_topic(self, request_messages, response_content):
-        text = "\n\n".join(
-            [message["content"] for message in request_messages]
-            + [response_content]
-        )
-
-        return self.get_topic_by_text(text)
-
     def get_topic_by_text(self, text):
         topics, _ = self.model.transform([text])
         topic = self.model.get_topic_info(topics[0])

--- a/tests/test_app.py
+++ b/tests/test_app.py
@@ -1,4 +1,5 @@
 import json
+import re
 
 from fastapi.testclient import TestClient
 
@@ -11,7 +12,7 @@ class TestTopicModel:
         return "TestTopic"
 
 
-def test_data_request():
+def test_chat_completion():
     write_api_mock = InfluxWriterMock()
     app.app.dependency_overrides[app.InfluxWriterAsync] = lambda: write_api_mock
     app.app.dependency_overrides[app.TopicModel] = lambda: TestTopicModel()
@@ -97,8 +98,296 @@ def test_data_request():
     ]
 
 
-def test_data_request_with_new_format():
+def test_chat_completion_list_content():
     write_api_mock = InfluxWriterMock()
+    app.app.dependency_overrides[app.InfluxWriterAsync] = lambda: write_api_mock
+    app.app.dependency_overrides[app.TopicModel] = lambda: TestTopicModel()
+
+    client = TestClient(app.app)
+    response = client.post(
+        "/data",
+        json=[
+            {
+                "message": json.dumps(
+                    {
+                        "apiType": "DialOpenAI",
+                        "chat": {"id": "chat-1"},
+                        "project": {"id": "PROJECT-KEY"},
+                        "user": {"id": "", "title": ""},
+                        "deployment": "gpt-4",
+                        "request": {
+                            "protocol": "HTTP/1.1",
+                            "method": "POST",
+                            "uri": "/openai/deployments/gpt-4/chat/completions?api-version=2023-03-15-preview",
+                            "time": "2023-08-16T19:42:39.997",
+                            "body": json.dumps(
+                                {
+                                    "messages": [
+                                        {
+                                            "role": "system",
+                                            "content": [
+                                                {
+                                                    "type": "text",
+                                                    "text": "act as a helpful assistant",
+                                                }
+                                            ],
+                                        },
+                                        {"role": "user", "content": "Hi!"},
+                                    ],
+                                    "model": "gpt-4",
+                                    "max_tokens": 2000,
+                                    "stream": True,
+                                    "n": 1,
+                                    "temperature": 0.0,
+                                }
+                            ),
+                        },
+                        "response": {
+                            "status": "200",
+                            "body": 'data: {"id":"chatcmpl-1","object":"chat.completion.chunk","created":1692214960,"model":"gpt-4","choices":[{"index":0,"delta":{"role":"assistant","content":"Hi"},"finish_reason":null}]}\n\ndata: {"id":"chatcmpl-1","object":"chat.completion.chunk","created":1692214960,"model":"gpt-4","choices":[{"index":0,"delta":{},"finish_reason":"stop"}],"usage":{"completion_tokens":189,"prompt_tokens":22,"total_tokens":211}}\n\ndata: [DONE]\n',
+                        },
+                    }
+                )
+            },
+        ],
+    )
+    assert response.status_code == 200
+    assert write_api_mock.points == [
+        'analytics,core_parent_span_id=undefined,core_span_id=undefined,deployment=gpt-4,execution_path=undefined,language=undefined,model=gpt-4,parent_deployment=undefined,project_id=PROJECT-KEY,response_id=chatcmpl-1,title=undefined,topic=TestTopic,trace_id=undefined,upstream=undefined chat_id="chat-1",completion_tokens=189i,deployment_price=0,number_request_messages=2i,price=0,prompt_tokens=22i,user_hash="undefined" 1692214959997000000',
+    ]
+
+
+def test_chat_completion_none_content():
+    write_api_mock = InfluxWriterMock()
+    app.app.dependency_overrides[app.InfluxWriterAsync] = lambda: write_api_mock
+    app.app.dependency_overrides[app.TopicModel] = lambda: TestTopicModel()
+
+    client = TestClient(app.app)
+    response = client.post(
+        "/data",
+        json=[
+            {
+                "message": json.dumps(
+                    {
+                        "apiType": "DialOpenAI",
+                        "chat": {"id": "chat-1"},
+                        "project": {"id": "PROJECT-KEY"},
+                        "user": {"id": "", "title": ""},
+                        "deployment": "gpt-4",
+                        "request": {
+                            "protocol": "HTTP/1.1",
+                            "method": "POST",
+                            "uri": "/openai/deployments/gpt-4/chat/completions?api-version=2023-03-15-preview",
+                            "time": "2023-08-16T19:42:39.997",
+                            "body": json.dumps(
+                                {
+                                    "messages": [
+                                        {
+                                            "role": "user",
+                                            "content": "what's the weather like?",
+                                        },
+                                        {
+                                            "role": "assistant",
+                                            "tool_calls": [
+                                                {
+                                                    "id": "xyz",
+                                                    "type": "function",
+                                                    "function": {
+                                                        "name": "get_weather",
+                                                        "arguments": {},
+                                                    },
+                                                }
+                                            ],
+                                        },
+                                        {
+                                            "role": "tool",
+                                            "id": "xyz",
+                                            "content": "It's sunny today.",
+                                        },
+                                        {
+                                            "role": "user",
+                                            "content": "2+3=?",
+                                        },
+                                    ],
+                                    "model": "gpt-4",
+                                    "max_tokens": 2000,
+                                    "stream": True,
+                                    "n": 1,
+                                    "temperature": 0.0,
+                                }
+                            ),
+                        },
+                        "response": {
+                            "status": "200",
+                            "body": 'data: {"id":"chatcmpl-1","object":"chat.completion.chunk","created":1692214960,"model":"gpt-4","choices":[{"index":0,"delta":{"role":"assistant","content":"Hi"},"finish_reason":null}]}\n\ndata: {"id":"chatcmpl-1","object":"chat.completion.chunk","created":1692214960,"model":"gpt-4","choices":[{"index":0,"delta":{},"finish_reason":"stop"}],"usage":{"completion_tokens":189,"prompt_tokens":22,"total_tokens":211}}\n\ndata: [DONE]\n',
+                        },
+                    }
+                )
+            },
+        ],
+    )
+    assert response.status_code == 200
+    assert write_api_mock.points == [
+        'analytics,core_parent_span_id=undefined,core_span_id=undefined,deployment=gpt-4,execution_path=undefined,language=undefined,model=gpt-4,parent_deployment=undefined,project_id=PROJECT-KEY,response_id=chatcmpl-1,title=undefined,topic=TestTopic,trace_id=undefined,upstream=undefined chat_id="chat-1",completion_tokens=189i,deployment_price=0,number_request_messages=2i,price=0,prompt_tokens=22i,user_hash="undefined" 1692214959997000000',
+    ]
+
+
+def test_embeddings_plain_text():
+    write_api_mock: app.InfluxWriterAsync = InfluxWriterMock()
+    app.app.dependency_overrides[app.InfluxWriterAsync] = lambda: write_api_mock
+    app.app.dependency_overrides[app.TopicModel] = lambda: TestTopicModel()
+
+    client = TestClient(app.app)
+    response = client.post(
+        "/data",
+        json=[
+            {
+                "message": json.dumps(
+                    {
+                        "apiType": "DialOpenAI",
+                        "chat": {"id": "chat-1"},
+                        "project": {"id": "PROJECT-KEY"},
+                        "user": {"id": "", "title": ""},
+                        "deployment": "text-embedding-3-small",
+                        "token_usage": {
+                            "completion_tokens": 0,
+                            "prompt_tokens": 2,
+                            "total_tokens": 2,
+                            "deployment_price": 0.001,
+                            "price": 0.001,
+                        },
+                        "parent_deployment": "assistant",
+                        "trace": {
+                            "trace_id": "5dca3d6ed5d22b6ab574f27a6ab5ec14",
+                            "core_span_id": "9ade2b6fef0a716d",
+                            "core_parent_span_id": "20e7e64715abbe97",
+                        },
+                        "execution_path": [None, "b", "c"],
+                        "request": {
+                            "protocol": "HTTP/1.1",
+                            "method": "POST",
+                            "uri": "/openai/deployments/text-embedding-3-small/embeddings?api-version=2023-03-15-preview",
+                            "time": "2023-08-16T19:42:39.997",
+                            "body": json.dumps({"input": ["fish", "cat"]}),
+                        },
+                        "response": {
+                            "status": "200",
+                            "body": json.dumps(
+                                {
+                                    "data": [
+                                        {
+                                            "embedding": [0.1, 0.2],
+                                            "index": 0,
+                                            "object": "embedding",
+                                        },
+                                        {
+                                            "embedding": [0.3, 0.4],
+                                            "index": 1,
+                                            "object": "embedding",
+                                        },
+                                    ],
+                                    "model": "text-embedding-3-small",
+                                    "object": "list",
+                                    "usage": {
+                                        "prompt_tokens": 2,
+                                        "total_tokens": 2,
+                                    },
+                                }
+                            ),
+                        },
+                    }
+                )
+            },
+        ],
+    )
+    assert response.status_code == 200
+    assert len(write_api_mock.points) == 1
+    assert re.match(
+        'analytics,core_parent_span_id=20e7e64715abbe97,core_span_id=9ade2b6fef0a716d,deployment=text-embedding-3-small,execution_path=undefined/b/c,language=undefined,model=text-embedding-3-small,parent_deployment=assistant,project_id=PROJECT-KEY,response_id=[^,]+,title=undefined,topic=TestTopic,trace_id=5dca3d6ed5d22b6ab574f27a6ab5ec14,upstream=undefined chat_id="chat-1",completion_tokens=0i,deployment_price=0.001,number_request_messages=2i,price=0.001,prompt_tokens=2i,user_hash="undefined" 1692214959997000000',
+        write_api_mock.points[0],
+    )
+
+
+def test_embeddings_tokens():
+    write_api_mock: app.InfluxWriterAsync = InfluxWriterMock()
+    app.app.dependency_overrides[app.InfluxWriterAsync] = lambda: write_api_mock
+    app.app.dependency_overrides[app.TopicModel] = lambda: TestTopicModel()
+
+    client = TestClient(app.app)
+    response = client.post(
+        "/data",
+        json=[
+            {
+                "message": json.dumps(
+                    {
+                        "apiType": "DialOpenAI",
+                        "chat": {"id": "chat-1"},
+                        "project": {"id": "PROJECT-KEY"},
+                        "user": {"id": "", "title": ""},
+                        "deployment": "text-embedding-3-small",
+                        "token_usage": {
+                            "completion_tokens": 0,
+                            "prompt_tokens": 2,
+                            "total_tokens": 2,
+                            "deployment_price": 0.001,
+                            "price": 0.001,
+                        },
+                        "parent_deployment": "assistant",
+                        "trace": {
+                            "trace_id": "5dca3d6ed5d22b6ab574f27a6ab5ec14",
+                            "core_span_id": "9ade2b6fef0a716d",
+                            "core_parent_span_id": "20e7e64715abbe97",
+                        },
+                        "execution_path": [None, "b", "c"],
+                        "request": {
+                            "protocol": "HTTP/1.1",
+                            "method": "POST",
+                            "uri": "/openai/deployments/text-embedding-3-small/embeddings?api-version=2023-03-15-preview",
+                            "time": "2023-08-16T19:42:39.997",
+                            "body": json.dumps(
+                                {"input": [[1, 3, 4, 5], [6, 7, 8, 9]]}
+                            ),
+                        },
+                        "response": {
+                            "status": "200",
+                            "body": json.dumps(
+                                {
+                                    "data": [
+                                        {
+                                            "embedding": [0.1, 0.2],
+                                            "index": 0,
+                                            "object": "embedding",
+                                        },
+                                        {
+                                            "embedding": [0.3, 0.4],
+                                            "index": 1,
+                                            "object": "embedding",
+                                        },
+                                    ],
+                                    "model": "text-embedding-3-small",
+                                    "object": "list",
+                                    "usage": {
+                                        "prompt_tokens": 2,
+                                        "total_tokens": 2,
+                                    },
+                                }
+                            ),
+                        },
+                    }
+                )
+            },
+        ],
+    )
+    assert response.status_code == 200
+    assert len(write_api_mock.points) == 1
+    assert re.match(
+        'analytics,core_parent_span_id=20e7e64715abbe97,core_span_id=9ade2b6fef0a716d,deployment=text-embedding-3-small,execution_path=undefined/b/c,language=undefined,model=text-embedding-3-small,parent_deployment=assistant,project_id=PROJECT-KEY,response_id=[^,]+,title=undefined,topic=TestTopic,trace_id=5dca3d6ed5d22b6ab574f27a6ab5ec14,upstream=undefined chat_id="chat-1",completion_tokens=0i,deployment_price=0.001,number_request_messages=2i,price=0.001,prompt_tokens=2i,user_hash="undefined" 1692214959997000000',
+        write_api_mock.points[0],
+    )
+
+
+def test_data_request_with_new_format():
+    write_api_mock: app.InfluxWriterAsync = InfluxWriterMock()
     app.app.dependency_overrides[app.InfluxWriterAsync] = lambda: write_api_mock
     app.app.dependency_overrides[app.TopicModel] = lambda: TestTopicModel()
 
@@ -206,7 +495,7 @@ def test_data_request_with_new_format():
     ]
 
 
-def test_rate_response_request():
+def test_rate_request():
     write_api_mock = InfluxWriterMock()
     app.app.dependency_overrides[app.InfluxWriterAsync] = lambda: write_api_mock
     app.app.dependency_overrides[app.TopicModel] = lambda: TestTopicModel()

--- a/tests/test_app.py
+++ b/tests/test_app.py
@@ -303,7 +303,7 @@ def test_embeddings_plain_text():
     assert response.status_code == 200
     assert len(write_api_mock.points) == 1
     assert re.match(
-        'analytics,core_parent_span_id=20e7e64715abbe97,core_span_id=9ade2b6fef0a716d,deployment=text-embedding-3-small,execution_path=undefined/b/c,language=undefined,model=text-embedding-3-small,parent_deployment=assistant,project_id=PROJECT-KEY,response_id=[^,]+,title=undefined,topic=TestTopic,trace_id=5dca3d6ed5d22b6ab574f27a6ab5ec14,upstream=undefined chat_id="chat-1",completion_tokens=0i,deployment_price=0.001,number_request_messages=2i,price=0.001,prompt_tokens=2i,user_hash="undefined" 1692214959997000000',
+        'analytics,core_parent_span_id=20e7e64715abbe97,core_span_id=9ade2b6fef0a716d,deployment=text-embedding-3-small,execution_path=undefined/b/c,language=undefined,model=text-embedding-3-small,parent_deployment=assistant,project_id=PROJECT-KEY,response_id=(.+?),title=undefined,topic=TestTopic,trace_id=5dca3d6ed5d22b6ab574f27a6ab5ec14,upstream=undefined chat_id="chat-1",completion_tokens=0i,deployment_price=0.001,number_request_messages=2i,price=0.001,prompt_tokens=2i,user_hash="undefined" 1692214959997000000',
         write_api_mock.points[0],
     )
 
@@ -381,7 +381,7 @@ def test_embeddings_tokens():
     assert response.status_code == 200
     assert len(write_api_mock.points) == 1
     assert re.match(
-        'analytics,core_parent_span_id=20e7e64715abbe97,core_span_id=9ade2b6fef0a716d,deployment=text-embedding-3-small,execution_path=undefined/b/c,language=undefined,model=text-embedding-3-small,parent_deployment=assistant,project_id=PROJECT-KEY,response_id=[^,]+,title=undefined,topic=TestTopic,trace_id=5dca3d6ed5d22b6ab574f27a6ab5ec14,upstream=undefined chat_id="chat-1",completion_tokens=0i,deployment_price=0.001,number_request_messages=2i,price=0.001,prompt_tokens=2i,user_hash="undefined" 1692214959997000000',
+        'analytics,core_parent_span_id=20e7e64715abbe97,core_span_id=9ade2b6fef0a716d,deployment=text-embedding-3-small,execution_path=undefined/b/c,language=undefined,model=text-embedding-3-small,parent_deployment=assistant,project_id=PROJECT-KEY,response_id=(.+?),title=undefined,topic=TestTopic,trace_id=5dca3d6ed5d22b6ab574f27a6ab5ec14,upstream=undefined chat_id="chat-1",completion_tokens=0i,deployment_price=0.001,number_request_messages=2i,price=0.001,prompt_tokens=2i,user_hash="undefined" 1692214959997000000',
         write_api_mock.points[0],
     )
 

--- a/tests/test_app.py
+++ b/tests/test_app.py
@@ -1,5 +1,4 @@
 import json
-from unittest.mock import Mock
 
 from fastapi.testclient import TestClient
 
@@ -7,13 +6,15 @@ import aidial_analytics_realtime.app as app
 from tests.influx_writer_mock import InfluxWriterMock
 
 
+class TestTopicModel:
+    def get_topic_by_text(self, text):
+        return "TestTopic"
+
+
 def test_data_request():
     write_api_mock = InfluxWriterMock()
     app.app.dependency_overrides[app.InfluxWriterAsync] = lambda: write_api_mock
-
-    topic_model = Mock()
-    topic_model.get_topic.return_value = "TestTopic"
-    app.app.dependency_overrides[app.TopicModel] = lambda: topic_model
+    app.app.dependency_overrides[app.TopicModel] = lambda: TestTopicModel()
 
     client = TestClient(app.app)
     response = client.post(
@@ -99,10 +100,7 @@ def test_data_request():
 def test_data_request_with_new_format():
     write_api_mock = InfluxWriterMock()
     app.app.dependency_overrides[app.InfluxWriterAsync] = lambda: write_api_mock
-
-    topic_model = Mock()
-    topic_model.get_topic.return_value = "TestTopic"
-    app.app.dependency_overrides[app.TopicModel] = lambda: topic_model
+    app.app.dependency_overrides[app.TopicModel] = lambda: TestTopicModel()
 
     client = TestClient(app.app)
     response = client.post(
@@ -211,10 +209,7 @@ def test_data_request_with_new_format():
 def test_rate_response_request():
     write_api_mock = InfluxWriterMock()
     app.app.dependency_overrides[app.InfluxWriterAsync] = lambda: write_api_mock
-
-    topic_model = Mock()
-    topic_model.get_topic.return_value = "TestTopic"
-    app.app.dependency_overrides[app.TopicModel] = lambda: topic_model
+    app.app.dependency_overrides[app.TopicModel] = lambda: TestTopicModel()
 
     client = TestClient(app.app)
     response = client.post(

--- a/tests/test_app.py
+++ b/tests/test_app.py
@@ -228,7 +228,7 @@ def test_chat_completion_none_content():
     )
     assert response.status_code == 200
     assert write_api_mock.points == [
-        'analytics,core_parent_span_id=undefined,core_span_id=undefined,deployment=gpt-4,execution_path=undefined,language=undefined,model=gpt-4,parent_deployment=undefined,project_id=PROJECT-KEY,response_id=chatcmpl-1,title=undefined,topic=TestTopic,trace_id=undefined,upstream=undefined chat_id="chat-1",completion_tokens=189i,deployment_price=0,number_request_messages=2i,price=0,prompt_tokens=22i,user_hash="undefined" 1692214959997000000',
+        'analytics,core_parent_span_id=undefined,core_span_id=undefined,deployment=gpt-4,execution_path=undefined,language=undefined,model=gpt-4,parent_deployment=undefined,project_id=PROJECT-KEY,response_id=chatcmpl-1,title=undefined,topic=TestTopic,trace_id=undefined,upstream=undefined chat_id="chat-1",completion_tokens=189i,deployment_price=0,number_request_messages=4i,price=0,prompt_tokens=22i,user_hash="undefined" 1692214959997000000',
     ]
 
 

--- a/tests/test_app.py
+++ b/tests/test_app.py
@@ -12,7 +12,7 @@ class TestTopicModel:
         return "TestTopic"
 
 
-def test_chat_completion():
+def test_chat_completion_plain_text():
     write_api_mock = InfluxWriterMock()
     app.app.dependency_overrides[app.InfluxWriterAsync] = lambda: write_api_mock
     app.app.dependency_overrides[app.TopicModel] = lambda: TestTopicModel()

